collector/src/main.rs

```diff
@@ -745,137 +686,196 @@ fn main() {
     }
 }
 
-fn main_result() -> anyhow::Result<i32> {
-    env_logger::init();
+#[derive(Debug, clap::Parser)]
+#[clap(about, version, author)]
+struct Cli {
+    #[clap(subcommand)]
+    command: Commands,
+}
 
-    let matches = clap_app!(rustc_perf_collector =>
-        (version: "0.1")
-        (author: "The Rust Compiler Team")
-        (about: "Collects Rust performance data")
+#[derive(Debug, clap::Args)]
+struct RustcOption {
+    /// The path to the local rustc to benchmark
+    // Not a `PathBuf` because it can be a file path *or* a `+`-prefixed
+    // toolchain name, and `PathBuf` doesn't work well for the latter.
+    rustc: String,
+}
 
-         // For each subcommand we list the mandatory arguments in the required
-         // order, followed by the options in alphabetical order.
+#[derive(Debug, clap::Args)]
+struct IdOption {
+    /// Identifier to associate benchmark results with
+    id: String,
+}
 
-        (@subcommand bench_local =>
-            (about: "Benchmarks a local rustc")
+#[derive(Debug, clap::Args)]
+struct LocalOptions {
+    /// Measure the build profiles in this comma-separated list
+    // This must be normalized via `ProfilerKind::expand_all()` before use.
+    #[clap(
+        long = "builds",
+        arg_enum,
+        multiple_values = true,
+        use_delimiter = true,
+        require_delimiter = true,
+        // Don't run rustdoc by default
+        default_value = "Check,Debug,Opt",
+    )]
+    profile_kinds: Vec<ProfileKind>,
+
+    /// The path to the local Cargo to use
+    #[clap(long, parse(from_os_str))]
+    cargo: Option<PathBuf>,
+
+    /// Exclude all benchmarks in this comma-separated list
+    #[clap(long)]
+    exclude: Option<String>,
+
+    /// Include only benchmarks in this comma-separated list
+    #[clap(long)]
+    include: Option<String>,
+
+    /// The path to the local rustdoc to benchmark
+    #[clap(long, parse(from_os_str))]
+    rustdoc: Option<PathBuf>,
+
+    /// Measure the scenarios in this comma-separated list
+    #[clap(
+        long = "runs",
+        arg_enum,
+        multiple_values = true,
+        use_delimiter = true,
+        require_delimiter = true,
+        default_value = "All"
+    )]
+    scenario_kinds: Vec<ScenarioKind>,
+}
 
-            // Mandatory arguments
-            (@arg RUSTC: +required +takes_value "The path to the local rustc to benchmark")
-            (@arg ID:    +required +takes_value "Identifier to associate benchmark results with")
+#[derive(Debug, clap::Args)]
+struct SelfProfileOption {
+    /// Collect self-profile data
+    #[clap(long = "self-profile")]
```

Comment:
I think you can use the inferred `long` (`#[clap(long)]`)

Reply:
I swear I had that originally and the option got named `--self_profile` instead of `--self-profile`. But I just tried it and it worked like you said. Huh.

Reply:
Oh... I just tried it with `ProfileLocal::out_dir` and for that one I ended up with `--out_dir` (with an underscore). So it seems to give different behaviour depending on whether you are in a delegated struct via `flatten`, yikes. Could that be a bug in clap?

I might keep the explicit names for now, because of this issue.

Reply:
Oh fun, I think you caught a bug.  I'm guessing we aren't defaulting the casing correctly for a struct-variant.

Reply:
Looks like the bug is that a `rename_all` on an enum is incorrectly applying to the variant structs
```rust
use clap::{Parser, Subcommand};

#[derive(Parser, Debug)]
#[clap(author, version, about, long_about = None)]
struct Cli {
    #[clap(long)]
    out_dir: Option<String>,

    #[clap(subcommand)]
    command: Commands,
}

#[derive(Subcommand, Debug)]
#[clap(rename_all = "snake_case")]
enum Commands {
    /// Adds files to myapp
    Add {
        #[clap(long)]
        out_dir: Option<String>,
    },
}

fn main() {
    let cli = Cli::parse();
    dbg!(cli);
}
```

Created https://github.com/clap-rs/clap/issues/3282